crates/agglayer-node/src/service/mod.rs

```diff
@@ -0,0 +1,390 @@
+use std::sync::Arc;
+
+use agglayer_config::{epoch::BlockClockConfig, Config, Epoch};
+use agglayer_storage::{
+    columns::latest_settled_certificate_per_network::SettledCertificate,
+    stores::{
+        DebugReader, DebugWriter, PendingCertificateReader, PendingCertificateWriter, StateReader,
+        StateWriter,
+    },
+};
+use agglayer_telemetry::KeyValue;
+use agglayer_types::{
+    Certificate, CertificateHeader, CertificateId, CertificateStatus, EpochConfiguration, Height,
+    NetworkId,
+};
+use ethers::{providers::Middleware, types::H256};
+use futures::future::try_join;
+use tokio::sync::mpsc;
+use tracing::{debug, error, info, instrument, trace};
+
+pub use self::error::{
+    CertificateRetrievalError, CertificateSubmissionError, SendTxError, TxStatusError,
+};
+use crate::{kernel::Kernel, signed_tx::SignedTx};
+
+pub mod error;
+
+/// The RPC agglayer service implementation.
+pub(crate) struct AgglayerService<Rpc, PendingStore, StateStore, DebugStore> {
+    kernel: Kernel<Rpc>,
+    certificate_sender: mpsc::Sender<(NetworkId, Height, CertificateId)>,
+    pending_store: Arc<PendingStore>,
+    state: Arc<StateStore>,
+    debug_store: Arc<DebugStore>,
+    config: Arc<Config>,
+}
+
+impl<Rpc, PendingStore, StateStore, DebugStore>
+    AgglayerService<Rpc, PendingStore, StateStore, DebugStore>
+{
+    /// Create an instance of the RPC agglayer service.
+    pub(crate) fn new(
+        kernel: Kernel<Rpc>,
+        certificate_sender: mpsc::Sender<(NetworkId, Height, CertificateId)>,
+        pending_store: Arc<PendingStore>,
+        state: Arc<StateStore>,
+        debug_store: Arc<DebugStore>,
+        config: Arc<Config>,
+    ) -> Self {
+        Self {
+            kernel,
+            certificate_sender,
+            pending_store,
+            state,
+            debug_store,
+            config,
+        }
+    }
+
+    /// Get access to the configuration.
+    pub fn config(&self) -> &Config {
+        &self.config
+    }
+}
+
+impl<Rpc, PendingStore, StateStore, DebugStore> Drop
+    for AgglayerService<Rpc, PendingStore, StateStore, DebugStore>
+{
+    fn drop(&mut self) {
+        info!("Shutting down the agglayer service");
+    }
+}
+
+impl<Rpc, PendingStore, StateStore, DebugStore>
+    AgglayerService<Rpc, PendingStore, StateStore, DebugStore>
+where
+    Rpc: Middleware + 'static,
+    PendingStore: PendingCertificateWriter + PendingCertificateReader + 'static,
+    StateStore: StateReader + StateWriter + 'static,
+    DebugStore: DebugReader + DebugWriter + 'static,
+{
+    #[instrument(skip(self, tx), fields(hash, rollup_id = tx.tx.rollup_id), level = "info")]
+    pub async fn send_tx(&self, tx: SignedTx) -> Result<H256, SendTxError<Rpc>> {
+        let hash = format!("{:?}", tx.hash());
+        tracing::Span::current().record("hash", &hash);
+
+        info!(
```

Comment:
Nit: we should probably use consistently structured logging format for all new log lines - discussed [here](https://github.com/agglayer/protocol-research/issues/59).

Reply:
This is taken from the original code in the `rpc` module. I did want to minimize behaviour changes in this PR (although some log lines may have changed), so would prefer this as a follow up.

Reply:
Agreed, we should do this consistently as a separate task.